### PR TITLE
In-Game Economy [Merge-Ready]

### PR DESCRIPTION
**Pull request by ghost**
This PR adds an economy to Space Station 13. At the beginning of the round, all players will receive a set amount of credits based on their job. These credits may be used for many things.

**ATMs** are now all around the station in the hallways and departments. These ATMs are installed by the Nanotrasen Bank of Credit and will accept any standard Nanotrasen identification card along with a PIN that comes with said card. ATMs are used to communicate with the central banking server as well as convert digital credit on an ID card to physical credits (and vice versa). An emag can be used on these to put them into maintenance mode, allowing you to bypass initial PIN prompts when inserting an ID.

**Bank servers** are new, unbuildable machines found at roundstart in the vault. These machines are immune to explosions and fire, and are linked to all station ATMs. Bank servers hold the data of all bank accounts and keep data relatively safe. They are protected by powerful network antiviruses, and only something powerful could hack into it...

This is not yet on the map and will be added to the map in an update following the merging of this one due to me forgetting to update my master before making this branch.

### CODE REVIEW

**Comment by Xhuis** [comment]
criticize my code remie-sama

File: code/game/objects/items/weapons/cards_ids.dm [hunk 1]
```diff
@@ -80,11 +80,33 @@
...
+	var/credits = 100 //Amount of dosh stored on card
+	var/pin = 1234 //The 4-digit PIN number used in ATMs
 	slot_flags = SLOT_ID
 
 	var/assignment = null
 	var/dorm = 0		// determines if this ID has claimed a dorm already
 
+/obj/item/weapon/card/id/New()
+	..()
+	sleep(10)
```
Comment (RemieRichards) [inline code/game/objects/items/weapons/cards_ids.dm]: why.

File: code/game/objects/items/weapons/cards_ids.dm [hunk 2]
```diff
@@ -80,11 +80,33 @@
...
+	pin = rand(0000,9999)
+	if(!ishuman(src.loc))
+		return
+	var/mob/living/carbon/human/H = src.loc
+	H << "<span class='notice'>The randomized PIN for your identification card this shift is [pin]. This can be changed at any ATM.</span>"
+	if(H.job == "Chief Medical Officer" || H.job == "Chief Engineer" || H.job == "Research Director" || H.job == "Head of Personnel" || H.job == "Head of Security")
```
Comment (RemieRichards) [inline code/game/objects/items/weapons/cards_ids.dm]: hacky.

```if(H.job in pay_bonus_jobs)``` please

File: code/game/objects/items/weapons/cards_ids.dm [hunk 3]
```diff
@@ -80,11 +80,33 @@
...
+		H << "<span class='info'>You have recieved a credit bonus to your ID card due to your position as a [H.job].</span>"
+		return
+	else if(H.job == "Quartermaster")
+		H << "<span class='info'>Being the head of commerce on [world.name], you have recieved a credit bonus to your ID card.</span>"
+		credits += 100
+		return
+	else if(H.job == "Captain")
```
Comment (RemieRichards) [inline code/game/objects/items/weapons/cards_ids.dm]: condense into the above.

**Comment by Incoming5643** [comment]
I still say there's no way our player base will want/humor/tolerate this

File: code/game/objects/items/weapons/cards_ids.dm [hunk 4]
```diff
@@ -80,11 +80,33 @@
...
+	pin = rand(0000,9999)
+	if(!ishuman(src.loc))
+		return
+	var/mob/living/carbon/human/H = src.loc
+	H << "<span class='notice'>The randomized PIN for your identification card this shift is [pin]. This can be changed at any ATM.</span>"
+	if(H.job == "Chief Medical Officer" || H.job == "Chief Engineer" || H.job == "Research Director" || H.job == "Head of Personnel" || H.job == "Head of Security")
+		H << "<span class='info'>You have recieved a credit bonus to your ID card due to your position as a [H.job].</span>"
+		return
+	else if(H.job == "Quartermaster")
+		H << "<span class='info'>Being the head of commerce on [world.name], you have recieved a credit bonus to your ID card.</span>"
```
Comment (RemieRichards) [inline code/game/objects/items/weapons/cards_ids.dm]: ```world.name``` should probably the ```station_name()``` proc

**Comment by ike709** [comment]
I tried porting bay capitalism to NT once but I cried after a while and gave up...

:+1:

**Comment by freerealestate** [comment]
The problem I've had with Bay and Goon's money system was that nothing being sold was worth buying, and everything was too cheap or you had too much money to spend. So not only did it not get used, you could clean out a vending machine easily if you wanted to, rendering it pointless.

This would have to be carefully balanced so things actually have the value they're meant to, but still allow them reasonable access to things that they should have (e.g. food).

I've been wanting this for awhile, so I sure hope it works out.

**Comment by daman997** [comment]
holler holler get $

**Comment by SconesC** [comment]
Does this mean we'll see Departmental Budgets?

**Comment by optimumtact** [comment]
Looks pointless but good effort anyway

**Comment by Dorsisdwarf** [comment]
QM is not a head!

:+1: make this work

**Comment by PKPenguin321** [comment]
as someone who hasn't really played outside /tg/, what exactly does this change do

**Comment by Miauw62** [comment]
oh come on i was going to do this.

File: code/game/objects/items/weapons/cards_ids.dm [hunk 5]
```diff
@@ -80,11 +80,33 @@
...
 	var/dorm = 0		// determines if this ID has claimed a dorm already
 
+/obj/item/weapon/card/id/New()
+	..()
+	sleep(10)
+	pin = rand(0000,9999)
+	if(!ishuman(src.loc))
+		return
+	var/mob/living/carbon/human/H = src.loc
+	H << "<span class='notice'>The randomized PIN for your identification card this shift is [pin]. This can be changed at any ATM.</span>"
```
Comment (Miauw62) [inline code/game/objects/items/weapons/cards_ids.dm]: add pin to notes

File: code/game/machinery/atm.dm [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+/obj/machinery/atm/attackby(obj/item/weapon/W, mob/user, params)
+	if(stat & (NOPOWER|BROKEN))
+		return
+	if(istype(W, /obj/item/weapon/card/id) && !id && !authed)
+		user.drop_item()
+		user.visible_message("<span class='notice'>[user] inserts [W] into [src].</span>", \
+							 "<span class='notice'>You slide the ID into [src].</span>")
+		W.loc = src
+		id = src
```
Comment (MrPerson) [inline code/game/machinery/atm.dm]: id = W, right?

**Comment by MrPerson** [comment]
Can we not do the PIN number crap? It's pointlessly secure. If you steal someone's ID, you should have their credits. Requiring a PIN just makes using the ID annoying.

**Comment by Czarcastic** [comment]
^this

**Comment by Iamgoofball** [comment]
Yeah, drop the PIN number stuff and just make it ID tied.

**Comment by Iamgoofball** [comment]
@freerealestate Money is mainly used for roleplay on goon and bay.

Like, yesterday on goon, I had some fun roleplay involving cash. It ended with me lighting the entire station budget on fire with a welder.

**Comment by WJohn** [comment]
This sounds neat, although there's an awful lot of systems that need to be redone to incorporate this at its minimum, and a lot more if you plan on making it actually worthwhile and interesting.

File: code/game/objects/items/weapons/cards_ids.dm (hunk 2 above)
Comment (Iamgoofball) [inline code/game/objects/items/weapons/cards_ids.dm]: if anything the text should also be defined on the job datum

**Comment by Ricotez** [comment]
:-1: 
why the hell would we need an economy? on those rare occasions where someone builds a shop, trade works just fine

**Comment by Iamgoofball** [comment]
@Ricotez 

Because it's fun.

**Comment by Steelpoint** [comment]
What does this bring to the game?

What game systems are changed to this new economy system? Is it just some sole ATM in the corner of arrivals?

Is this change actually going to be a positive one?

Do I have to look up Bay and Goon wiki's to figure out exactly what this PR does to a fine T?

**Comment by freerealestate** [comment]
If you read the code, all it does for now (as the PR isn't finished) is add an ATM and give IDs debit card functionality.

**Comment by Steelpoint** [comment]
Thank you for explaining something that should have been in the OP to begin with.

**Comment by WJohn** [comment]
This is more infrastructure than anything: It paves the way for new economy-centered systems and additions in the future.

Think of being able to both buy and sell almost any item on the station at cargo bay for a quick buck. 
What about being able to buy and sell station power? That could come in extremely handy if you have an excess or are lacking power.
Perhaps in the future if we ever add a trader system (either entirely interacted with through trading consoles or actual walking talking NPCs) you'll suddenly be able to use space bucks to trade with them instead of the really arbitrary "cargo point". Makes a lot more sense.

**Comment by Steelpoint** [comment]
My main concern with a economy system is how its implemented, I don't want it to become a case of us now forcing people to buy food from vending machines.

I'll be interested in seeing how this is explained and discussed. But so long as it does not become something that can heavily influence rounds, it should be a optional thing to the side.

**Comment by Miauw62** [comment]
"my main concern with an economy system is that it will be used"
#wow #whoa

**Comment by Steelpoint** [comment]
I said how its implemented, again I'm simply wary of this being a hamfisted addition that forces people to buy things they prior did not have to.

**Comment by Ricotez** [comment]
I see absolutely no point in adding this, the station is a top secret reseach facility built by Nanotrasen, not a trade hub or a space city.

**Comment by Chnkr** [comment]
Please add this, it adds more ~fun~ and ~immursions~ into the game, so i can now go and steal the station budget, and light it on fire.

**Comment by WJohn** [comment]
I'm pretty sure nanotrasen is a super greedy and super assholish corporate nightmare. It'd make plenty of sense to throw in tons of advertisements and force you to pay for bunches of commodities (like snacks and drinks).

Department specific things like YouTools, medivends and so on should remain free, of course.

**Comment by xxalpha** [comment]
Bad and not necessary. Goes against the play style of tg servers. :-1:

**Comment by Xhuis** [comment]
I'm gonna address some concerns here before I continue working on this.

* __This is a top secret space station and should not have an economy.__

This is headcanon. There is no official lore on the /tg/ servers, unlike other servers like Bay and /vg/, so if you want to think it's top secret you can, but there's also people who will vehemently tell you that it's a well-known station.

* __This goes against the playstyle on /tg/ servers.__

What? I'd like some more explanation on this, please.

* __What exactly does this bring to the game?__

To begin with, it will spawn every crew member with a certain amount of credits at the beginning of the round. These credits can be manipulated by inserting your ID into an ATM found in certain areas. In addition, each crew member's ID card will have a randomly-generated number between 0001 and 9999. This number is the card's PIN number and must be authenticated at an ATM before credits can be changed.

The ATM can be used to change PIN and accept physical money to convert it into digital money on the ID (or vice versa - digital money to physical). This PR will serve as the framework for implementation of an economy, and future PRs will implement this system in further ways.

**Comment by xxalpha** [comment]
If you played on sybil you'd know this is pointless. You will have players ignore your hardwork and call shekel collectors shitters for charging for services that have always been free (Because this is a crew on a ship and all work torwards a common goal).

**Comment by Steelpoint** [comment]
Technically the lore supports SS13 being a top secret space station.

**Comment by freerealestate** [comment]
Some of the randomised station names include "top secret" in them. Not
exactly concrete evidence, but it does indicate past developers have
considered it top secret. I don't know why its secretiveness matters
though. It could trade under a façade, posing as a different type of
station. It can't be that secret if the corporation's enemies are
constantly finding their way on board.

**Comment by Steelpoint** [comment]
No, I mean the actual lore that we've had for a very long time, found on the wiki.

**Comment by freerealestate** [comment]
Yeah, but who wrote it?

**Comment by Steelpoint** [comment]
We'll never know simply because its been there for so long, and so rarely edited, that there's no record of who wrote it originally.

**Comment by freerealestate** [comment]
Also, the gameplay should inform the lore, not the other way around.

**Comment by Steelpoint** [comment]
Yeah but that does not detract from the facts.

Either way, I like the idea of a departmental budget, but I just don't want anything to go overboard, keep the economy mostly regulated to Cargo and ordering equipment.

**Comment by ike709** [comment]
Let the battle for capitalism between Miauw and Xhuis begin.

File: code/game/machinery/atm.dm [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+		W.loc = src
+		id = W
+		AttemptToInterface(user)
+	return
+
+/obj/machinery/atm/attack_hand(mob/user)
+	AttemptToInterface(user)
+
+/obj/machinery/atm/proc/AttemptToInterface(mob/user)
+	/*if(user.key == "Miauw")
```
Comment (Chnkr) [inline code/game/machinery/atm.dm]: You should uncomment this.

Comment (Miauw62) [inline code/game/machinery/atm.dm]: nice meme

Comment (Xhuis) [inline code/game/machinery/atm.dm]: dang the secret is out

**Comment by Summoners** [comment]
:+1: I would like any economy added to the game, even if the servers don't want to use it they could simply disable it on the main server, but it would be nice to have in the code for future uses.

File: code/game/machinery/atm.dm [hunk 8]
```diff
@@ -0,0 +1,122 @@
+/obj/machinery/atm
+	name = "automated transfer machine"
+	desc = "A Nanotrasen Bank of Credit ATM, used for trasnferring funds to and from ID cards."
```
Comment (Chnkr) [inline code/game/machinery/atm.dm]: trasnferring?

Comment (Xhuis) [inline code/game/machinery/atm.dm]: totally

File: code/game/machinery/atm.dm [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+/obj/machinery/atm/power_change()
+	if(stat & BROKEN)
+		return
+	else
+		if(powered())
+			stat &= ~NOPOWER
+		else
+			icon_state = "dorm_off"
+			stat |= NOPOWER
+			EjectId() //If we run out of power, eject an ID if we have one
```
Comment (Chnkr) [inline code/game/machinery/atm.dm]: You have no power to eject an ID? You could do the same as the hop console does, require an simple screwdriver to unscrew and remove the ID

**Comment by Xhuis** [comment]
ATMs put on the map, plus a banking server in the vault and a small system for handling bank accounts.

**Comment by Xhuis** [comment]
fuck

**Comment by Xhuis** [comment]
ok i think i got it right this time finally

**Comment by Xhuis** [comment]
welp

**Comment by Xhuis** [comment]
it hath been unfucked

**Comment by Xhuis** [comment]
This is merge-ready. Once it gets merged I can add it to the maps since it won't conflict anymore.

**Comment by pudl** [comment]
@Xhuis I way prefer this system to Miauw's, I hope it gets merged! :)

**Comment by Miauw62** [comment]
>merge-ready

wot

**Comment by Miauw62** [comment]
also why do you even prefer this to my system @pudl

File: code/game/machinery/economy.dm [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+			return EjectId(user)
+		else
+			say("Authentication accepted. Welcome, [id.registered_name].")
+			authed = 1
+	Interface(user)
+
+/obj/machinery/atm/proc/Interface(mob/user)
+	if(!user.canUseTopic(src)) return
+	var/list/options = list("Change card PIN #", "Create New Account", "Access Banking Account", "Exit")
+	switch(input(user, "You have [id.credits] credits on your ID.", "ATM Interface", null) in options)
```
Comment (Miauw62) [inline code/game/machinery/economy.dm]: this is awful.
make an actual UI pls

Comment (freerealestate) [inline code/game/machinery/economy.dm]: I don't think this will get merged just because it was finished sooner,
Miauw. I expect the maintainers will weigh the pros and cons of each as
well as player opinion before deciding.

Comment (Miauw62) [inline code/game/machinery/economy.dm]: im just saying that using input() as an UI is pretty terrible

Comment (Xhuis) [inline code/game/machinery/economy.dm]: i can't into ui

**Comment by Rinoka** [comment]
Opinion: The banking server makes very little sense from any perspective. Why would Nanotransen put all the banking data on a station doomed to get utterly wrekt, lose power, and float away?

I'd say it'd be best to change it to a transceiver-type thing, keeping all its normal functions but making it so it can be destroyed or deconstructed, and all data on it is stored at Centcomm and transferred to it at regular intervals. Or maybe just fuck that and put the broadcaster on Centcomm instead.

Also, it'd be best to give the money a name that holds no value. IE, Nanocredits, instead of space money or something that implies actual value. Lets everyone preserve headcanons about how the station works without playing around too much.

**Comment by Xhuis** [comment]
Turning the server into a transceiver or some huge router of sorts would be easy enough.

File: code/game/machinery/economy.dm (hunk 10 above)
Comment (Iamgoofball) [inline code/game/machinery/economy.dm]: Thanks for volunteering to code a UI miauw

**Comment by Chnkr** [comment]
You could ask @nullquery to be nice, and use his recently merged UI system to build a ~nice~ UI.

File: code/game/machinery/economy.dm (hunk 10 above)
Comment (duncathan) [inline code/game/machinery/economy.dm]: he did
for his own version

**Comment by optimumtact** [comment]
Or you could learn how to use it AND THEN DO IT YOURSELF

**Comment by nullquery** [comment]
If you need help getting the UI to work (or if you'd rather that I take a look myself) let me know and I'll take action.

**Comment by pudl** [comment]
@Miauw62 Honestly, I dislike both of them a lot when compared to Goon and Bay's systems, but I think that we need economy either way, and I think that Xhuis' system is nearly perfect, I just really detest banking servers and the entire idea of it. I wish it had never been brought up.

**Comment by Incoming5643** [comment]
need is a strong word

**Comment by psykzz** [comment]
Im loving the competition keep it up!

 :+1: to economy,
also :+1: to having it in now and off if that is what is preferred.

**Comment by Xhuis** [comment]
@pudl What method would you prefer other than banking servers?

**Comment by nullquery** [comment]
Interface is done. It may need some tweaking before it goes live, but it's functional.

**Comment by Xhuis** [comment]
HTML interface added to ATMs - all credit goes entirely to @nullquery for creating and implementing a UI that isn't just input() chains!

**Comment by pudl** [comment]
@Xhuis I prefer having a PIN, instead of it simply being a matter of "steal id and done". I prefer the bank servers not be destructible, because that's following in the path of telecomms, which is a system that always gets sabotaged and is rarely fixed after it's ruined. However, I don't think ATMs should be emaggable.

**Comment by Rinoka** [comment]
@pudl Initial plan (and as far as I can tell, it wasn't changed) was to have the PIN number. The bank servers (satellite) should be destructible or at least depowerable because your logic isn't correct. Telecomms are brought online if the round's not already gone to crap, and I think that anything else should follow the same rule.

It seems like you want a system that can't be easily fucked up, when that's the antithesis of anything in SS13.

**Comment by Xhuis** [comment]
@pudl In the new version, accounts are not linked to IDs. You just need to insert an ID into the ATM to authenticate the fact that you are a valid crewmember.

**Comment by nullquery** [comment]
There's two versions, ID and account-based storage.

Benefit of ID-based storage is that you have a physical card that holds your credits. But it can be stolen. And then someone could steal your money.

At the moment it's not possible to deposit/withdraw cash from your ID. IMO it should be possible to do this, withdrawing money should result in paper money the same way that occurs in Bay, and depositing money back into the ATM should allow you to put it back on your card (I guess this would work by bluespace, teleporting the physical money from/to a banking station somewhere on the map).

Benefit of account-based storage is that your money is not on your ID and more difficult to steal by thieves. But those in command may be able to use their privileges to get to your money by manipulating your accounts (through manipulation of the banking server). But at the moment it's not possible to manipulate accounts without knowing the PIN, IMO there should be a banking server monitor that grants unrestricted access to accounts and a global transfer log.

**Comment by pudl** [comment]
Why not just have it so that the account and your ID are two separate entities. You can put money into an ATM and transfer said money to your bank account (where it's hypothetically inaccessible to anyone without a pin) where your payroll is stored, or put it onto your ID, which makes it so swiping your ID on shit actually transfers money. That way emagging an ATM can give you the money from the ID, but bank accounts are still untouchable.

**Comment by Dorsisdwarf** [comment]
Just put it in the vault, we've been trying to get people to care about vault security for months/years

**Comment by Miauw62** [comment]
@pudl this is basically bays system though. and bays system only really works because of the high rp there.
also my banking servers aren't like tcomms, because they require no setup at all. if your scientists are competent enough to repair the cloner, they can easily build a new or extra banking server.
in my case, money is also not lost if the banking server dies, all it keeps is the logs.

File: code/game/machinery/economy.dm [hunk 11]
```diff
@@ -0,0 +1,406 @@
...
+	src.ejectID()
+	atms_in_world.Remove(src)
+	return ..()
+
+/obj/machinery/atm/proc/validatePIN(pin)
+	pin = text2num(pin)
+	return pin >= 1000 && pin <= 9999
+
+/obj/machinery/atm/Topic(href, href_list[], datum/html_interface_client/hclient)
+	if (istype(hclient))
```
Comment (Miauw62) [inline code/game/machinery/economy.dm]: >unsanitized Topic()

you can literally steal the credits of anybody that puts their card into an ATM and route them to your account by exploiting the hrefs here.

**Comment by Miauw62** [comment]
>nullquery makes UI for xhuis

:/

**Comment by Xhuis** [comment]
v('-')v

File: code/game/machinery/economy.dm (hunk 11 above)
Comment (nullquery) [inline code/game/machinery/economy.dm]: No, you can't -- the html interface module deals with securing the topic calls in the background. So long as you check the hclient you can be assured that the call is legitimate (player is allowed to use the device and has a window open).

**Comment by Aranclanos** [comment]
:-1: for not being finished, the money is useless

**Comment by Steelpoint** [comment]
I really wish there was a single thread for this.

Overall I don't think as it stands a economic system would do anything at all for /tg/station aside from sitting there idle, come up with actual usages for a economic system, run it up a design document, post it on the forums and then code the system instead of having a unfinished system and promising you'll finish it later.

Considering Shadowling was never finished I don't have huge faith here.

**Comment by Rinoka** [comment]
It's just a framework, but I'd rather be handed the tools to add stuff using it instead of having it be implemented in a way nobody likes in-game (ie, paying for drinks) and having it axed.

**Comment by nullquery** [comment]
Some brainstorming (without other people, they can really mess your ideas up :laughing:):

**Banking servers**
These are nice, but there should be 2 of them AT LEAST. One on the station and one at Central Command.

Each banking server handles transactions and manages local accounts. These accounts can be registered at will, the same way they can be handled now.

You can assign as much credits as you want to any account (as if you had var edit :grin:) if you have direct write access to the bank server.

Initially there should only be 1 bank server per z-level. ATM's should seek out this bank server upon creation. If additional servers are constructed then you can create ATM's which are slaved to those servers.

**ATM's**
These connect to the active bank server for the current z-level and allow transactions to be performed against its accounts.

Presently ID cards have a certain amount of credits assigned to them (that is, directly assigned to the object). This is wrong IMO. ID cards should not have any credits assigned to them, but should instead be linked to a bank account on the local bank server.

That way the ID card can't be physically manipulated to modify the credits, but instead you can only manipulate the account it references (by copying a fingerprint from another ID).

When players get ID's they should also get an accompanying bank account that stores the credits.

**Bank accounts**
Bank accounts should be loose entities. Using an ATM should not force you to use your ID, instead you should be able to connect to a bank account directly by specifying the name and unique PIN for the bank account.

But what about the bank account that's linked to your ID? Yes, it's linked, but that's only for ease-of-use: it allows you to automatically login to your bank account when you insert your ID without having to specify the PIN, and it allows you to perform transactions using that linked bank account when you use your ID on vending machines and cash registers.

Your ID card then becomes your wallet, but you can create a second bank account to deposit funds into to avoid theft.

Since ID's can be used to directly authenticate without specifying a PIN they should be kept secure. The HoP and Captain may sometimes need your ID but they should have direct write access to the local bank server already so they could mess with accounts directly and move your credits to another account.

**Transactions between bank servers**
Transactions between bank servers don't care about accounts. You still need to specify an account to route the funds to, but this is a formality and can actually be skipped.

The only thing two bank servers know about each other is their worth. They store this information about each other, so the CC bank server knows the total amount of credits on the station bank server. The station bank server is slaved to the CC bank server, so it doesn't care about the amount of credits -- it'll accept any amount it sends, assuming it to be correct. This is the default configuration, but with direct write access to a bank server you can change how it perceives the other servers.

**tl;dr**

What I'm proposing is to design the system in such a way that the owners of the bank server (be it CentCom/NT or the Captain of the station) have full control over the credits that flow through them, but keep things secure so that you can't steal all the credits.

**Comment by Miauw62** [comment]
you could read my thread/PR, which contains a lot of your ideas.

**Comment by Aranclanos** [comment]
try to not post ideas here, github comments are already bad enough. Imagine if everyone was posting their two page long walltext of ideas that nobody wants to read
Also your ideas add no real value to the currency, issue that both economy prs have

**Comment by Xhuis** [comment]
bippity bop this is still merge-ready

**Comment by Chnkr** [comment]
@Xhuis can i steal this?

**Comment by Xhuis** [comment]
for another server? uhh sure

**Comment by Aranclanos** [comment]
again, this is not finished and I don't think it should be merged

**Comment by Iamgoofball** [comment]
it's finished

**Comment by Xhuis** [comment]
as i've said, the datedness of this branch prevents me from adding things to the maps, and i want to get the framework in before i add any more to the server

**Comment by freerealestate** [comment]
The money doesn't have value. You should think specifically of what crewmembers will spend their money on, and how they'll do it, and add that, and then the system you've coded will have purpose.

**Comment by Iamgoofball** [comment]
@freerealestate 

you're missing the point

the point is so we have the system in place so that everyone can contribute features instead of all the eggs in one basket which results in an old outdated pr

**Comment by Aranclanos** [comment]
is there even a plan yet to give value to the currency

**Comment by Xhuis** [comment]
Me and goof had a pretty in-depth discussion a while back. Plans are to add it to Cargo, create digitized cash registers in service areas, replace mining points and vending machines, and also things like stock markets and gambling.

**Comment by freerealestate** [comment]
So ideally you want people to ask money for things they have previously given away for free (e.g. chef's food, botany's plants, research's prototypes). But where's their incentive to? What can they then do with the money? Sure, they could go buy food from the chef or plants from botany (in this hypothetical example) but there's still no real incentive to accumulate money instead of just handing over stuff for free. They don't actually need the money.

If there was something worthwhile that money granted you that you couldn't get any other way (e.g. for free, by theft, or creating it yourself) then there would be an incentive to charge for services and care about money accumulation. An idea might be off-station traders who you can't rob or otherwise physically access who sell items that a player might value (unique clothes, weapons (this'll open up a can of shit), chems, tools, materials, etc.).

 Another thing I just thought of for actually adding value to the items output by a department is by making the materials needed to create the items cost money. Hence every department has to balance a budget, and can no longer give away stuff for free. I know you kind of already have this planned with departmental budgets (at least I think that's this pull) and changing cargo supply points into dollars, which I guess would be earned back in turn by charging the budgets of the departments they're supplying.

Just some stuff to think about. But I don't think this is getting merged until you show people why money will be worth using.

**Comment by Xhuis** [comment]
Well, the players' feedback is in and majority votes that an economy will not be good for the server. I'm sorry to say that this is being closed. I said when I first uploaded this PR that if the players didn't want it, it wouldn't happen, and I'm not going back on my word.

**Comment by phil235** [comment]
awww :disappointed:

**Comment by Aranclanos** [comment]
the economy can work,, it can be fun, but you missed the most important part of it, giving the money a real value

**Comment by freerealestate** [comment]
If you came up with some actual examples of incentivising (not forcing)
players to use money and made a new poll, it might not turn out so
negatively.

**Comment by Xhuis** [comment]
Polls played little part in the decision - personally asking people how they felt about an economy in general did.

**Comment by Iamgoofball** [comment]
@freerealestate fuck off commie, you know the only way to get people to use it is to force at least SOME use of money to get people used to it

**Comment by Rinoka** [comment]
People are rejecting this merge because, I think, they think something will go wrong, or people will be forced to use the system, or yadda yadda. This assumes quite a bit about how it'll be implemented and people sure are assuming they won't have power of veto for individual PRs.

I want this to be merged, because most of everyone's complaints are "I cannot *believe* this is what you'd do with the system" or "You need to say what you'd do with the system". If it ends up nobody finds a way to use it at the end of the day (aka after giving it some time) without earning the ire of the community, then get it reverted.

Yes, Xhuis is the one implementing the system, but that does not make them responsible for knowing what can be done with it - especially when "what can we DO with it" is outside the scope of this PR, and they SAID that. There's a lot of interesting, non-forceful stuff that can be done to incentivize the system, but Goof is right - forcing it into use if only a little bit would be pretty much the only way to get people familiar with it and make them use it in the longer run.

**Comment by freerealestate** [comment]
Forcing is fine, provided accumulating money is fun (i.e. you have
worthwhile things to spend it on).

**Comment by nullquery** [comment]
IMO features like this should be modular, put this into "modules/economy" and write it in such a way that if the code files aren't included that the economy stuff isn't added to the server.

That way people who want to use it can use it, people who don't want to use it can exclude it.

**Comment by optimumtact** [comment]
that leaves us with a problem of code rot however, because stuff that isn't on the tgstation servers tends to be unmaintained and we don't often see downstream patches for it
